### PR TITLE
Comparison and Boolean Operations

### PR DESCRIPTION
**Pull request by apparentlymart**
This set of commits introduces new operators to support comparison and boolean operations. The new operators are as follows:

* Equality: `==` and `!=`
* Numeric comparison: `<`, `>`, `<=` and `>=`
* Boolean operations: `&&`, `||` and unary `!`
* Ternary conditional: `... ? ... : ...`

This is a logical (:grimacing:) extension of work started in #26.

I went with "C-style" operator punctuation because I think this has been the syntax most prolifically emulated by other programming languages and so users are likely to be familiar with it from at least one of C's descendants. Additionally, HCL (in which HIL is most commonly embedded) is using "C-inspired" brace-delimited blocks. (perhaps it's more accurate to attribute this syntax to [B](https://en.wikipedia.org/wiki/B_(programming_language)), but hey...)

My primary motivation for working on this is to make a further step towards the use-cases discussed in hashicorp/terraform#1604.

A pattern has emerged for "faking" conditional resources in Terraform using variables mapping to counts, as follows:

```hcl
variable "env_decider" {
  production = 1
  staging    = 0
}

resource "example" "foo" {
  # ...
  count = "${var.env_decider[var.environment]}"
}
```

Including this changeset in Terraform would immediately "pave the cowpath" here and introduce a more succinct way to express that idea:

```hcl
resource "example" "foo" {
  # ...
  count = "${var.environment == "production" ? 1 : 0}"
}
```

In the discussion on hashicorp/terraform#1604 there is also some consensus around a more "intentional" mechanism for conditionally including resources, which Terraform could support via some additional features building on the boolean operations added in these commits:

```hcl
# will not work without further changes to Terraform, but would build on these changes
resource "example" "foo" {
  # ...
  if = "${var.environment == "production"}"
  # (other suggestions for this attribute name have been "when", "onlyif", ...
}
```

---

Due to the same type system limitations that befell full implementation of #34, for this initial changeset the `... ? ... : ...` operator is constrained to not work on array or map types. In principle it could support direct usage of variables of these types in a similar way to how indexing is supported, but since this seems to be a rather marginal case I decided to omit this complexity right now until either the type system gets improved to represent collection element types or until we find a specific motivating use-case.

### CODE REVIEW

File: ast/conditional.go [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+	FalseExpr Node
+	Posx      Pos
+}
+
+// Accept passes the given visitor to the child nodes in this order:
+// CondExpr, TrueExpr, FalseExpr. It then finally passes itself to the visitor.
+func (n *Conditional) Accept(v Visitor) Node {
+	n.CondExpr = n.CondExpr.Accept(v)
+	n.TrueExpr = n.TrueExpr.Accept(v)
+	n.FalseExpr = n.FalseExpr.Accept(v)
```
Comment (mitchellh) [inline ast/conditional.go]: I'll leave this for the future (not this PR), but I think we can do something clever here somehow to make it so we don't have to fully evaluate both sides (lazy evaluate).

File: check_types.go [hunk 2]
```diff
@@ -177,6 +192,116 @@ func (tc *typeCheckArithmetic) TypeCheck(v *TypeCheck) (ast.Node, error) {
...
+		}
+	}
+
+	// Only ints and floats can have the <, >, <= and >= operators applied
+	switch tc.n.Op {
+	case ast.ArithmeticOpEqual, ast.ArithmeticOpNotEqual:
+		// anything goes
+	default:
+		switch compareType {
+		case ast.TypeFloat, ast.TypeInt:
```
Comment (mitchellh) [inline check_types.go]: I think strings could make sense here, just copying behavior of Go spec's string comparison.

File: ast/conditional.go (hunk 1 above)
Comment (apparentlymart) [inline ast/conditional.go]: Ahh yes, that's true. I touched on that briefly and then moved on considering it to just be a marginal performance thing, but now that you mention it I suppose some specific things like Terraform's `file(...)` function could benefit from being short-circuited both here and in the `&&` and `||` operators as a future improvement, so people can do stuff like:

```
something = "${var.data_file != "" ? file(var.data_file) : "default_something"}"
```

**Comment by VladRassokhin** [comment]
Great feature!
Do you have plan to include that to Terraform 0.8? 
And if yes when would be next rc/release?

**Pull request merged**